### ISSUE TITLE
inline parsing algorithm, spec conflicts with implementation

### ISSUE THREAD

**Issue by rsc** [commonmark/CommonMark#686]
The parsing strategy described in https://spec.commonmark.org/0.30/#appendix-a-parsing-strategy does not appear to reflect the actual implementations. 

Specifically, the section "An algorithm for parsing nested emphasis and links" would suggest that other kinds of inlines are handled separately, either before or after. But it appears to be impossible for that to be the case, at least for code spans.

1. If code spans were handled before the "An algorithm", then `` [foo](`url`) `` would not parse as a link, yet it does in the dingus. 
2. If code spans were handled after the "An algorithm", then `` `[foo](url)` `` would not parse as a code span, yet it does in the dingus.

It must therefore be the case that code spans are handled _during_ the "An algorithm", in some way that is not described.
That is, the algorithm conflicts with the implementation.

---

The spec also seems to conflicts with the implementation. It says:

> Code span backticks have higher precedence than any other inline constructs except HTML tags and autolinks.

And yet in the dingus `` [foo](`url`) `` is a link, not a code span with text around it.

---

There is a long discussion about backticks in link parens in #439 but it never reached any conclusion. But the outcome of #439 is independent of my main point, which is that the implementation and the algorithm+spec are out of sync.

**Comment by wooorm** [comment]
Given that every parser (on babelmark3) sees that as a link, rather than honing the “precedence”, I personally feel changing one sentence in a spec to reflect reality is the easiest, compared to trying to change every parser.
Personally ``[hello](`foo`)`` looks more like a link than code.
@rsc, as the OP, what do you think about these cases and their most sensical output?

---

One more complex example: `[text](<svg:rect>)`; the `<svg:rect>` is both valid “HTML”, a valid autolink, and a valid destination for the link. (similar cases: `<xml:lang/>` andr `<svg:circle{...props}>`)

**Comment by rsc** [comment]
The current left-to-right scan that takes things one token at a time seems to me the only viable possibility, given that different syntaxes can end up breaking up into different token sequences. It's not like an arithmetic expression grammar where you can tokenize in one pass, without looking at the operators, and then apply operator precedence in a different pass. That is, these are not really operators in the same way. So I don't think it's tenable to say that code spans really have higher precedence than link-closing tokens. Precedence is just the wrong concept. 

Changing the rule would make this a code span, not a link:

    [hello](foo "`quotes'")`

Yet the current spec is clear that this is an HTML tag, not a code span:

    x <a foo="`quotes'">`

The fact that the latter is not a code span means that you can't split out the code spans in a separate pass. The code span extractor would have to be at least aware of which backquotes were inside HTML tags, and that means doing the full HTML tag parser. And then probably autolinks too in that pass. 

It seems much clearer and cleaner to do it all in one pass, with only strong/emphasis application left to a second pass, instead of trying to build a sequence of passes that would almost certainly step on each other's feet.

The sentence

> Code span backticks have higher precedence than any other inline constructs except HTML tags and autolinks.

is almost true. It could be made true by saying

> Code span backticks have higher precedence than any other inline constructs except HTML tags, autolinks, and link closing ] sequences.

but at that point there's nothing with lower precedence that might conflict, so the sentence is essentially vacuous. And the complex case at the end of @woorm's comment just above this one is another place where notions of precedence would get very tricky.

The left-to-right in the first sentence of section 6 seems to be the key, both to the current implementations and to a simple mental model of what happens.

**Comment by jgm** [comment]
> but at that point there's nothing with lower precedence that might conflict, so the sentence is essentially vacuous.

I'm not sure I understand.
```
This is not *`emphasized*`.
```
Here you see that code takes precedence over emphasis markers.
```
This is not [`a link]`.

[`a link]: http://example.com
```
Here you see that code takes precedence over link markers.
The only exception is what happens in inline links, where the `](...)` part ignores what would otherwise be raw HTML or code.

**Comment by rsc** [comment]
You are right about the emphasis being lower precedence than code spans - I was not entirely precise about that. It remains true that emphasis is lower precedence than the others. My point is that the others are essentially all equal precedence.

> Here you see that code takes precedence over link markers.

That's not how I see this example.

It is not that code takes precedence over link markers as that code and link closing markers have the same precedence and the left-to-right rule is what disambiguates. Either code or link closing can "ignore" the other. The one that is leftmost wins.

    This is [a link](`x)`.

    This is [a code span: `](x`).

It is important to view link framing as two different markers - the `[` and the `]...` - and not just one, so that leftmost is evaluated correctly when a `]...` is involved.

**Comment by jgm** [comment]
> Either code or link closing can "ignore" the other.

This is correct about the link closers, yes (either the `](..)` in inline links or the `][..]` in explicit reference links).  But not about link openers, or the closing bracket in a link label in a reference link.

```
% cmark
[a link`][]`

[a link`]: foo
^D
<p>[a link<code>][]</code></p>
```
This is what I was thinking about with the remark about precedence.

Anyway, I want to agree with you that the remark about precedence is not quite right. However, there is something in that vicinity that is not vacuous that needs to be said.

**Comment by rsc** [comment]
I see what you mean. Perhaps the statement to make would be:

> Code spans, HTML tags, autolinks, link labels, link destinations, and link titles all have equal precedence and are disambiguated by parsing the input left-to-right.
